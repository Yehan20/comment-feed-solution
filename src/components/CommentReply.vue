<script setup lang="ts">

import { PxMessageReply, BiArrowUpCircle, BiArrowDownCircle } from "oh-vue-icons/icons";
import { addIcons } from 'oh-vue-icons';
import TextEditor from "./TextEditor.vue";
import { onMounted, onUpdated, ref } from "vue";
import { usecommentFeedStore } from "../store/commentFeedStore"
import { helperHasNestedMessages } from "../utils/messageHelper";
import BaseModal from "./BaseModal.vue";
import CommentVoteButtonGroup from "./CommentVoteButtonGroup.vue";


// Icom Method
addIcons(PxMessageReply, BiArrowDownCircle, BiArrowUpCircle);

// Props
const props = defineProps<{
    messageInfo: {
        id: number,
        upvoted: boolean,
        downvoted: boolean,
        parentId: number | null
        userName: string
    }
}>()


// Store
const commentFeedStore = usecommentFeedStore();


// refs
const canReplyComment = ref(false);

const hasNestedThreads = ref(helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id))

const showCommentThread = ref(helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id))// main rest


const showModel = ref(false);

// Local varaible

let message = 0;

onMounted(() => {
    //  console.log('mount comment',canReplyComment.value);
})


//Methods
const toggleReply = () => {
    canReplyComment.value = !canReplyComment.value
}

// Expland and Collapse Nested messages 
const handleCollapseThread = () => {
    showCommentThread.value = !showCommentThread.value
}

// Collapse the Messages in the level of the nevly added message 
const autoCollapseCommentThread = () => {

    if (showCommentThread.value) {
        showCommentThread.value = false
    }

}
const deleteCurrentMessage = () => {
    canReplyComment.value = showModel.value = false
}

const handleCancel = () => {

    // if message is empty   remove  the message if not show the model 
    (message) ? showModel.value = true :
        canReplyComment.value = !canReplyComment.value

}

// Updated Life Cycle

onUpdated(() => {
    hasNestedThreads.value = helperHasNestedMessages(props.messageInfo.parentId, commentFeedStore.commentFeed, props.messageInfo.id)
})

</script>
<template>
    <div class="comment__item__reply">

        <div class="comment__buttons__wrapper">
            <div class="bg-white rounded">

                <CommentVoteButtonGroup 
                 @@downVote="commentFeedStore.downvoteComment(props.messageInfo.id)"
                 @@upVote="commentFeedStore.upvoteComment(props.messageInfo.id)"
                 :upvoted="props.messageInfo.upvoted" 
                 :downvoted="props.messageInfo.downvoted"/>

            </div>


            <button class="btn" @click="handleCancel" v-if="canReplyComment"
                :title="canReplyComment ? 'Click to Cancel' : 'Click to Reply'">
                Cancel X
            </button>

            <button class="btn" title="Click to Reply" v-else @click="canReplyComment = true">
                Reply <v-icon scale=1 name="px-message-reply"></v-icon>
            </button>


            <button class="btn" v-if="hasNestedThreads" @click="handleCollapseThread"
                :title="showCommentThread ? 'Click to expand feed' : 'Click to hide feed'">
                {{ showCommentThread ? 'Expand +' : 'Hide -' }}


            </button>
        </div>


        <template v-if="canReplyComment">
            <TextEditor :parentId="messageInfo.id" :userName="messageInfo.userName ?? ''"
                @@autoCollapseNest="autoCollapseCommentThread" @@hideEditor="toggleReply"
                @@checkMessageLength="message = $event" :showDiscussion="false" />
        </template>



    </div>

    <template v-if="!showCommentThread">
        <slot />
    </template>

    <BaseModal title="Discard comment?"
        description="You have a comment in progress, are you sure you want to discard it? " v-if="showModel"
        @@close=" showModel = !showModel" @@delete="deleteCurrentMessage" />



</template>

<style scoped lang="scss">
@use '../assets/scss/variables.scss' as v;



.comment__item__reply {


    p {
        margin: 0;
        font-size: 13px;
        color: v.$darkGray;
        font-weight: 500;
    }

    .comment__buttons__wrapper {
        display: flex;
        gap: 10px;
        margin-bottom: 5px;
        button{
        
                text-transform: lowercase;
                border-radius: 20px;
                padding-left: 10px;
                padding-right: 10px;
   
                &.btn__upvote {
   
                   padding: 5px;
                   margin-right: 0;
               }
       
               &.btn__downvote {
                   padding: 5px;
                   margin-right: 0;
       
               }
          
               &.btn__upvote.selected {
                   color: v.$brand-color !important;
               }
       
               &.btn__downvote.selected {
                   color: v.$downvote-clr !important;
               }
       
               &:hover {
                   opacity: 0.7;
               }     
        }
      
    }

}
</style>
