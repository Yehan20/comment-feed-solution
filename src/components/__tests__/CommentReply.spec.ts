import { describe, it, expect, beforeEach, afterAll, beforeAll } from "vitest";
import { mount } from "@vue/test-utils";
import { createPinia, setActivePinia } from "pinia";
import CommentReply from "../CommentReply.vue";

import { nextTick } from "vue";



describe('CommentReply', () => {

   
    beforeEach(() => {
        setActivePinia(createPinia());
    });


    // Mounting Test
    it("Component Should Render Correctly", () => {

        const wrapper = mount(CommentReply, {
           props:{
            messageInfo:{
                 id:2,
                 parentId:1,
                 downvoted:false,
                 upvoted:false,
                 userName:"Yehan"
            }
           }

        })

        expect(wrapper.html()).toMatchSnapshot()

    })


        // Check Emit When Cancel the Model
    it('emits a message when the upvote button is clicked', async () => {

        const wrapper = mount(CommentReply, {
          props: {
            messageInfo: {
              id: 3,
              userName: "abc",
              parentId: 1,
              upvoted: false,
              downvoted: false,
            }
          }
        });
    
    

  });



})


