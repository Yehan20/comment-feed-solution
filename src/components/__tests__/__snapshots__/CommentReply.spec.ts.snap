// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentReply > Component Should Render Correctly 1`] = `
"<div data-v-7d6b8569="" class="comment__item__reply">
  <div data-v-7d6b8569="" class="comment__buttons__wrapper">
    <div data-v-7d6b8569="" class="bg-white rounded"><button data-v-1090a549="" class="btn btn__upvote" title="Click to Upvote">
        <v-icon data-v-1090a549="" scale="1.2" name="bi-arrow-up-circle"></v-icon>
      </button><button data-v-1090a549="" class="btn btn__downvote" title="Click to DownVote">
        <v-icon data-v-1090a549="" scale="1.2" name="bi-arrow-down-circle"></v-icon>
      </button></div><button data-v-7d6b8569="" class="btn reply__btn" title="Click to Reply"> Reply <v-icon data-v-7d6b8569="" scale="1" name="px-message-reply"></v-icon></button>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
<!--v-if-->"
`;
